Add `OperationSpaceController` to docs and tests and implement corresponding action/action_cfg classes

ozhanozen opened the pull request:
# Description

This PR adds the `OperationSpaceController` to the docs and provides some tests for its parametric features. Moreover, it implements the corresponding `OperationSpaceControllerAction` and `OperationSpaceControllerActionCfg` classes so they can be used with manager-based environments.

Fixes #873 

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist

- [ ] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [ ] I have added my name to the `CONTRIBUTORS.md` or my name already exists there
``

Dhoeller19 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/actions/task_space_actions.py:
```diff
@@ -190,3 +191,169 @@ def _compute_frame_jacobian(self):
+            self._joint_ids = slice(None)
+
+        # create the operation space controller
+        self._ik_controller = OperationSpaceController( #FIXME Change to correct arguments
```
Please fix comment

Dhoeller19 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/actions/task_space_actions.py:
```diff
@@ -190,3 +191,169 @@ def _compute_frame_jacobian(self):
+        else:
+            joint_pos_des = joint_pos.clone()
+        # set the joint position command
+        # self._asset.set_joint_position_target(joint_pos_des, self._joint_ids)
```
No action is applied right now

ozhanozen replied:
I have implemented the action class

ozhanozen commented:
@Dhoeller19, I have created this PR to replace #879 but haven't worked much on it yet. It is far from complete at the moment. I am currently working on it and will ping once it reaches a state worth reviewing. Sorry for the confusion.

jtigue-bdai commented:
Converted to a draft util ready for full review. @ozhanozen i will continue to work with you on this branch. Thanks for doing this!

ozhanozen commented:
@jtigue-bdai, no problem. It takes a bit of time to complete, though. I have some updates and questions regarding the progress.

I have added tests for cases where the target is abs/rel pos/pose, and w/wo inertial/gravity compensations. I took the differential IK controller test as a reference: if the robot ee reaches these targets well, the tests pass. The current tests pass successfully. The remaining tests to be implemented are:

- [ ] Force target (any idea what would be a test that makes sense here?)
- [ ] Hybrid target with selection matrix
- [ ] Variable stiffness (I will try varying the values within a reasonable range, for the sole purpose of testing but without any other function)

After these tests, the following steps would be:

- [ ] Implementing `OperationSpaceControllerAction` and `OperationSpaceControllerActionCfg`, which could basically be based on the test code.
- [ ] Adding OPC to the documentation.

Please let me know what you think and if you have any feedback. 

Btw, I noticed that sim.clear() inside the tearDown() makes my tests hang, if LIVESTREAM is **not** 2 and started. This is the same for other tests, not only this controller. I am working with dockerized version, which might be relevant. Do you have any idea why this might be the case?

ozhanozen commented:
@Mayankm96, I also wanted to ping you here as you wrote the original OPC implementations. 

I have made some changes to the implementation and wanted to ensure that these are accurate and that I haven't missed something that might break the functionality somewhere else. Mainly:
- I observed some problems with the batched matrix dimensions being incompatible with multiplications, so I changed them.
- The desired ee pos/rot/force was previously updated within the compute() function. However, this made the target unreachable when defined relative (e.g., the desired pose = ee pose + command was applied all the time). I moved this part to set_command() so the target is constant during the operation.
- Some minor naming changes to be consistent with differential ik controller.
- Separated OperationSpaceControllerCfg and OperationSpaceController to different files.

Please let me know if these changes are ok for you, or if you have any objections or feedback.

Moreover, there are some points regarding the current OPC implementation I was thinking about, which might be worth to consider for adding as a feature/modification:

1. Currently the target is assumed to have constant position/pose. We can generalize this by allowing inputting reference velocity/acceleration such that `des_ee_acc = p_gains * (pos-pos_ref) + d_gains * (vel-vel_ref) + acc_ref`. I believe this would be relevant as we one want that the robot follows a moving target.
2. We have the selection matrix to choose which axes we want to control the movement and which axes we want to apply the force. However, the axes are wrt the earth frame. In many cases, one might want that these axes are defined wrt the end-effector (e.g., we want to apply force in ee z-axis regardless of ee pose). We can achieve this by transforming the selection matrix using the rotation matrix between ee and the robot base. Maybe we can provide this as a parameter to the controller_cfg? (e.g., control_axes_frame: Literal["base", "ee"] )
3. There are scale coefficients for position and orientation commands. I see such coefficients are typically used within the corresponding Action implementations for other controllers. If these have the same purpose, does it make sense to remove them to be moved to future Action implementations?
4. The current implementation does not contain compensation for the velocity-dependent disturbances: the Coriolis/centrifugal forces and the velocity-induced acceleration due to the mapping of coordinates (i.e., `J^{dot}*q^{dot}`). Does it make sense to compensate for these? I believe it is easy to obtain the first one from the sim, but for the second element, we need the time derivative of the Jacobian, which I don't know if it is possible to obtain from the sim directly.
5. While the EE reaches the target successfully, the robot floats within its null space if a redundant robot is used. Shall we include some basic secondary targets (e.g., minimizing energy, maximizing manipulability, etc) to deal with this redundancy? 

I believe that some of these points might require significant work and make the implementation more complex. However, I think OPC might be beneficial for sim2real, and it is not a bad idea to have a complete implementation now. We could create another PR for some of these, though, as it goes beyond the scope of this PR.

What do you think about these points?

jtigue-bdai commented:
Thanks @ozhanozen, things are looking pretty good. I will give them a review.

As for you questions:
- Force target: Maybe spawn the robot in a pose and have it press into a plane with a desired force and check steady state torques (minus gravity torques) and see if they match `tau=J'w`
- Hybrid target: Similar to force target but have a desired normal force and a desired tangential position.
- Varying stiffness, I think that is reasonable. Fully validating exact stiffnesses may not be worth the effort can probably drive the arm into a desired position past a wall a couple and make sure quasi statically the forces go up for higher stiffnesss.
- As for the Action implementation do you mean an example or writing the actual action code? 
- The OPC is added into the documentation by adding the lines to [docs/source/api/lab/omni.isaac.lab.controllers.rst](https://github.com/isaac-sim/IsaacLab/pull/913/files#diff-f27d7b2f10ed397ea1868a31d6e16aa1360eea34415e81ce870d24ef06e5b54d) that are already done. The docs are autogenerated by docstrings. We might clean up the docstrings along the way though. in addition we an add a standalon example and make a how-to or tutorial addition to the documentation.

I will try running your tests on my machine. I also use the docker workflow. I will let you know what I find.

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/controllers/operational_space.py:
```diff
@@ -177,6 +116,14 @@ def num_actions(self) -> int:
     Operations.
     """
 
+    def reset(self, env_ids: torch.Tensor = None):
+        """Reset the internals.
+
+        Args:
+            env_ids: The environment indices to reset. If None, then all environments are reset.
+        """
+        pass
+
```
Might be good to reset desired_ee_pos, desired_ee_rot, and desired_ee_force

ozhanozen commented:
> Thanks @ozhanozen, things are looking pretty good. I will give them a review.
> 
> As for you questions:
> 
> * Force target: Maybe spawn the robot in a pose and have it press into a plane with a desired force and check steady state torques (minus gravity torques) and see if they match `tau=J'w`
> * Hybrid target: Similar to force target but have a desired normal force and a desired tangential position.
> * Varying stiffness, I think that is reasonable. Fully validating exact stiffnesses may not be worth the effort can probably drive the arm into a desired position past a wall a couple and make sure quasi statically the forces go up for higher stiffnesss.
> * As for the Action implementation do you mean an example or writing the actual action code?
> * The OPC is added into the documentation by adding the lines to [docs/source/api/lab/omni.isaac.lab.controllers.rst](https://github.com/isaac-sim/IsaacLab/pull/913/files#diff-f27d7b2f10ed397ea1868a31d6e16aa1360eea34415e81ce870d24ef06e5b54d) that are already done. The docs are autogenerated by docstrings. We might clean up the docstrings along the way though. in addition we an add a standalon example and make a how-to or tutorial addition to the documentation.
> 
> I will try running your tests on my machine. I also use the docker workflow. I will let you know what I find.

Thanks @jtigue-bdai for the suggestions, I will try to create test cases as you suggested.

Regarding the actions, I intended to implement the actual `Action` and `ActionCfg` code so that OPC is ready to be used with the manager-based environments. For this, I tried to write the test code in a way its functions could be re-used for the action implementation later.

ozhanozen commented:
> * Force target: Maybe spawn the robot in a pose and have it press into a plane with a desired force and check steady state torques (minus gravity torques) and see if they match `tau=J'w`
> * Hybrid target: Similar to force target but have a desired normal force and a desired tangential position.

Hi @jtigue-bdai, I have finished writing all the test cases but deviated a bit from what we discussed:

- For the force control, I didn't see any way to measure interaction torques at the joints, and thought comparing applied torques with the desired wrench using` tau=J'w` didn't make sense. This is exactly how the joint torques are calculated in the open-loop case, and the equivalence is satisfied all the time. What I did instead was to put walls around the robot and attach contact sensors to these walls. The robot EE crashes into these walls; then I check if the first part of the desired wrench (i.e., forces) equals the measured contact force. In general, this works, but it is not very precise since you cannot easily make the robot stop at the contact point while using only force control, and any movement affects the contact force. Therefore, I have tuned the wall positions, gains, etc, as much as I could, and defined more "relaxed" thresholds for checking the force equivalence. If you change these parameters, it is possible that the robot goes somewhere else and the test fails, but this is, I believe, how the real robot would work as well, so it is expected.

- For the hybrid control, I basically applied force control in only one axis pressing the wall, and I applied position control in the tangential directions, as you suggested. This time, the robot reaches a steady state pose more reliably. However, it is still not 100% robust as the force control can make the robot EE stuck if the path to the target pose is not simple or if the robot rotates in a bad way within its null space. Again, I tuned the tests so they pass, but they are a bit fragile whenever there is a force control component involved.

I will now proceed with writing the Action/ActionCfg classes. Feel free to provide any input for me to integrate.

Hi @Mayankm96, did you have a chance to look at the five improvement points I mentioned above? @jtigue-bdai, do you have any comments on these? I have thought about one more improvement, which could help with these force control issues I faced:

6. Let's say I would like robot EE to go to the table surface and apply force on the z-axis while I position-control the other axes. If the initial pose of the robot EE is too far from the target contact point, this sometimes creates problems as I mentioned; hence, I would normally first do a pure pose control until I reach the vicinity of the contact point, where I would switch to force control. I propose we include a new command specific to the force control (i.e., a binary or a float to fade smoothly) that allows us to switch/fade from motion control to force control for the axes the user wants to do eventually do force control. This way, not only can people use hybrid/force control in a more "controlled" manner, but also an RL policy could learn the optimal way to transition from motion to force control at the contact point. What do you think?

jtigue-bdai commented:
@ozhanozen thanks for all you hard work. I will take a look and get back with reviews.

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/controllers/operational_space.py:
```diff
@@ -185,43 +134,83 @@ def reset_idx(self, robot_ids: torch.Tensor = None):
         """Set target end-effector pose or force command.
 
         Args:
             command: The target end-effector pose or force command.
```
Can you expand this docstring to denote how the different impedance modes affect command shape and contents?

ozhanozen replied:
I have edited the docstring as the following:

```
 """Set the task-space targets and impedance parameters.

        Args:
            command (torch.Tensor): A concatenated tensor of shape ('num_envs', 'action_dim') containing task-space
                targets (i.e., pose/wrench) and impedance parameters.
            current_ee_pose (torch.Tensor, optional): Current end-effector pose of shape ('num_envs', 7), containing position
                and quaternion (w, x, y, z). Required for relative commands. Defaults to None.

        Format:
            Task-space targets, ordered according to 'command_types':

                Absolute pose: shape ('num_envs', 7), containing position and quaternion (w, x, y, z).
                Relative pose: shape ('num_envs', 6), containing delta position and rotation in axis-angle form.
                Absolute wrench: shape ('num_envs', 6), containing force and torque.

            Impedance parameters: stiffness for 'variable_kp', or stiffness, followed by damping ratio for 'variable':

                Stiffness: shape ('num_envs', 6)
                Damping ratio: shape ('num_envs', 6)

        """
```

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/controllers/operational_space.py:
```diff
@@ -185,43 +134,83 @@ def reset_idx(self, robot_ids: torch.Tensor = None):
+    def set_command(self, command: torch.Tensor, ee_pose: torch.Tensor | None = None):
         """Set target end-effector pose or force command.
 
         Args:
```
please add ee_pose to args in docstring with shape info

jtigue-bdai replied:
might also be good to add `current` or similar to the variable name so that it doesn't get confused with target

ozhanozen replied:
I have changed it to current_ee_pose (please see comment above).

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/controllers/operational_space.py:
```diff
@@ -185,43 +134,83 @@ def reset_idx(self, robot_ids: torch.Tensor = None):
             )
         # impedance mode
         if self.cfg.impedance_mode == "fixed":
             # joint positions
             self._task_space_target[:] = command
```
if this joint positions, then I think it would be better to name this `_joint_space_target`

ozhanozen replied:
It shouldnt be joint positions but rather task space targets. I have changed changed all accordingly.

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/controllers/operational_space.py:
```diff
@@ -185,43 +134,83 @@ def reset_idx(self, robot_ids: torch.Tensor = None):
+                target @= self._position_command_scale
+                # compute targets
+                self.desired_ee_pos = ee_pose[:, :3] + target
+                self.desired_ee_rot = ee_pose[:, 3:]
```
Does this get ignored when calculating the joint commands (i.e. opening up the null space) or does it just assume you want to keep the same orientation?

ozhanozen replied:
No, it is not ignored; you are right to point it out. It assumes you want to keep the same orientation, which one could also achieve with `pose_abs` or `pose_rel`. 

If one wants the orientation to be ignored, the correct way to do this is to set `motion_control_axes = (x, x, x, 0, 0, 0) `. In this case, do we actually need `position_abs` and `position_rel` (or even possibly `orientation_abs` and `orientation_rel`)? These could be all generalized.

Why don't we get rid of these options and simply encourage users to use `motion_control_axes` and `wrench_control_axes`? This would simplify the code a bit as well. The only catch might be the user needs to always provide ee_pose all the time, even if the orientation is not cared.

Alternatively, I can modify `motion_control_axes` depending on the command_type.

jtigue-bdai replied:
Yeah If we can get rid of the position specific ones and just use the motion_control_axes. That way the ee_pose is only needed for the pose_rel case and the pose_abs should be good as is.

ozhanozen replied:
I have removed them.

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/controllers/operational_space.py:
```diff
@@ -185,43 +134,83 @@ def reset_idx(self, robot_ids: torch.Tensor = None):
+                    raise ValueError("End-effector pose is required for 'position_abs' command.")
+                # compute targets
+                self.desired_ee_pos = target
+                self.desired_ee_rot = ee_pose[:, 3:]
```
Similar question as in position_rel

ozhanozen replied:
These are now removed (please see the comment above).

jtigue-bdai commented on source/extensions/omni.isaac.lab/test/controllers/test_operation_space.py:
```diff
@@ -0,0 +1,759 @@
+        )
+        ee_pose_b = torch.cat([ee_pos_b, ee_quat_b], dim=-1)
+
+        # Compute the current velocity of the end-effector
+        ee_vel_w = robot.data.body_vel_w[:, ee_frame_idx, :]  # Extract end-effector velocity in the world frame
+        root_vel_w = robot.data.root_vel_w  # Extract root velocity in the world frame
+        relative_vel_w = ee_vel_w - root_vel_w  # Compute the relative velocity in the world frame
+        ee_lin_vel_b = quat_rotate_inverse(robot.data.root_quat_w, relative_vel_w[:, 0:3])  # From world to root frame
+        ee_ang_vel_b = quat_rotate_inverse(robot.data.root_quat_w, relative_vel_w[:, 3:6])
+        ee_vel_b = torch.cat([ee_lin_vel_b, ee_ang_vel_b], dim=-1)
```
NOTE here that the linear velocity that you get is not necessarily at the frame of root/ee_frame. the output is at the COM of each of those RigidBodies.

jtigue-bdai replied:
Since the COM of the ee is likely at or near the link frame this isn't really an issue. Just wanted you to be aware of it.

ozhanozen replied:
It is a good point, and I believe it would be nice to correct this. Do you know how to do this? I do not know how to get COM frames.

Btw, I saw a similar problem with the Jacobian here #967, I guess the same problem applies for OPC. Hence, I took the solution there and adapted opc test code with the following:
```

# obtain dynamics related quantities from simulation
ee_jacobi_idx = ee_frame_idx - 1
jacobian_w = robot.root_physx_view.get_jacobians()[:, ee_jacobi_idx, :, arm_joint_ids]
mass_matrix = robot.root_physx_view.get_mass_matrices()[:, arm_joint_ids, :][:, :, arm_joint_ids]
gravity = robot.root_physx_view.get_generalized_gravity_forces()[:, arm_joint_ids]
# Convert the Jacobian from world to robot base frame
jacobian_b = jacobian_w.clone()
root_rot_matrix = matrix_from_quat(quat_inv(robot.data.root_state_w[:, 3:7]))
jacobian_b[:, :3, :] = torch.bmm(root_rot_matrix, jacobian_b[:, :3, :])
jacobian_b[:, 3:, :] = torch.bmm(root_rot_matrix, jacobian_b[:, 3:, :])
```

ozhanozen replied:
Ah, I have just seen #966; we can wait for that to be merged for the velocities.

jtigue-bdai reviewed (commented):
In general things are looking good.

ozhanozen commented on source/extensions/omni.isaac.lab/omni/isaac/lab/controllers/operational_space.py:
```diff
@@ -177,6 +116,14 @@ def num_actions(self) -> int:
+            env_ids: The environment indices to reset. If None, then all environments are reset.
+        """
+        pass
+
```
I have changed to:

```
def reset(self):
        """Reset the internals."""
        self.desired_ee_pos = None
        self.desired_ee_rot = None
        self.desired_ee_wrench = None
```

I removed `env_ids`, is it needed?

ozhanozen commented:
> > @Mayankm96, I also wanted to ping you here as you wrote the original OPC implementations.
> > I have made some changes to the implementation and wanted to ensure that these are accurate and that I haven't missed something that might break the functionality somewhere else. Mainly:
> > 
> > * I observed some problems with the batched matrix dimensions being incompatible with multiplications, so I changed them.
> > * The desired ee pos/rot/force was previously updated within the compute() function. However, this made the target unreachable when defined relative (e.g., the desired pose = ee pose + command was applied all the time). I moved this part to set_command() so the target is constant during the operation.
> > * Some minor naming changes to be consistent with differential ik controller.
> > * Separated OperationSpaceControllerCfg and OperationSpaceController to different files.
> > 
> > Please let me know if these changes are ok for you, or if you have any objections or feedback.
> > Moreover, there are some points regarding the current OPC implementation I was thinking about, which might be worth to consider for adding as a feature/modification:
> > 
> > 1. Currently the target is assumed to have constant position/pose. We can generalize this by allowing inputting reference velocity/acceleration such that `des_ee_acc = p_gains * (pos-pos_ref) + d_gains * (vel-vel_ref) + acc_ref`. I believe this would be relevant as we one want that the robot follows a moving target.
> > 2. We have the selection matrix to choose which axes we want to control the movement and which axes we want to apply the force. However, the axes are wrt the earth frame. In many cases, one might want that these axes are defined wrt the end-effector (e.g., we want to apply force in ee z-axis regardless of ee pose). We can achieve this by transforming the selection matrix using the rotation matrix between ee and the robot base. Maybe we can provide this as a parameter to the controller_cfg? (e.g., control_axes_frame: Literal["base", "ee"] )
> > 3. There are scale coefficients for position and orientation commands. I see such coefficients are typically used within the corresponding Action implementations for other controllers. If these have the same purpose, does it make sense to remove them to be moved to future Action implementations?
> > 4. The current implementation does not contain compensation for the velocity-dependent disturbances: the Coriolis/centrifugal forces and the velocity-induced acceleration due to the mapping of coordinates (i.e., `J^{dot}*q^{dot}`). Does it make sense to compensate for these? I believe it is easy to obtain the first one from the sim, but for the second element, we need the time derivative of the Jacobian, which I don't know if it is possible to obtain from the sim directly.
> > 5. While the EE reaches the target successfully, the robot floats within its null space if a redundant robot is used. Shall we include some basic secondary targets (e.g., minimizing energy, maximizing manipulability, etc) to deal with this redundancy?
> > 
> > I believe that some of these points might require significant work and make the implementation more complex. However, I think OPC might be beneficial for sim2real, and it is not a bad idea to have a complete implementation now. We could create another PR for some of these, though, as it goes beyond the scope of this PR.
> > What do you think about these points?
> 
> 1. I think for this feature its a reasonable thing to add, but lets hold off for a further PR. We are already adding quite a bit here.
> 2. I think this would be good to add especially for relative based targets (pose_rel, position_rel, and even force).
> 3. I think that is valid. You can move that to the action implementation.
> 4. Lets hold off on that for now, but I think it would be pretty beneficial later especially for faster motions.
> 5. An easy way to handle this would be to add some joint level damping in the actuators.
> 6. That would be useful, but we may want to save that for a further PR.

Thanks a lot for the answers.

The 3 is done. I will keep 1,4,6 for future PR. 2, I will try after `Action` implementations.

For 5, adding joint level damping might work for the tests but it is eventually undesired for sim2real as this damping is not available in the real robot. Also, somehow it is useful to have some control over the redundancy. What about I work on this separately and open a PR later if I have an elegant solution?

jtigue-bdai commented:
> For 5, adding joint level damping might work for the tests but it is eventually undesired for sim2real as this damping is not available in the real robot. Also, somehow it is useful to have some control over the redundancy. What about I work on this separately and open a PR later if I have an elegant solution?

Yeah that sounds good.

ozhanozen commented:
Hi @jtigue-bdai,

I wanted to give an update:
- I have finished implementing the `OperationalSpaceControllerAction` for the integration of OSC with manager-based environments. As an example, I have added a variation of the reach environment with minimal changes, [opc_env_cfg.py](source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/manipulation/reach/config/franka/opc_env_cfg.py), that allows learning/testing with OSC. Of course, not all combination of OSC parameters makes sense in this environment, since I have inherited from the existing reach environment config. Nevertheless, current parameters allow learning to reach, and it provides an example of OSC Action usage (you can use `"Isaac-Reach-Franka-OSC-v0" `and `"Isaac-Reach-Franka-Play-OSC-v0"`).
- There was a challenging issue regarding the contact sensors. If the OSC parameters are set in a way that a closed loop force feedback is necessary, I create a `ContactSensor` that belongs to the `OperationalSpaceControllerAction`. However, since the `ContactSensorCfg` is not part of the simulation scene directly, I couldn't make the simulation automatically handle the sensor (e.g., its initialization, reset, update, etc). Hence, I tried to manually handle the necessary `ContactSensor` operations (e.g., calling  `self._contact_sensor._initialize_impl()`). It seems to work, but it is not an elegant solution. It would be better if you could check it once more. Is there a better way to integrate the `ContactSensor`? (e.g., making it a part of the articulation). I didn't see any other examples.

There are three more steps missing from my side for this PR:
- Rotating the motion/force selection matrixes if the user wants to define them according to the end-effector (point 2 from the discussion before). I will work on this now, but it should be feasible.
- A small tutorial, either similar to [this](docs/source/tutorials/05_controllers/run_diff_ik.rst), or more focused on the Action implementation. What do you think?
- Whenever #966 is merged, I will recheck the test cases and tune them if necessary.

Please let me know if you have any feedback or reviews. We are almost there :)

jtigue-bdai commented:
Thanks @ozhanozen. I will review the code as it is currently and when #966 is merged I will let you know.

ozhanozen commented:
@jtigue-bdai, I have finished the following:

- Added the feature to define an arbitrary frame (i.e., task reference frame) according to which all the targets could be defined. For example, if one wants to apply a force perpendicular to a tilted wall surface, it is easier to define the force targets w.r.t. a frame that is on the wall. All the relevant matrices/vectors (motion/force selection matrices, relative rotation vector, stiffness values, etc) are transformed according to this task reference frame now. I believe this would be useful: the recent Forge paper followed a similar approach and got promising results. It works well according to my tests, but I haven't included this in the Action class, as this requires tracking a frame external to the robot asset. I will check how to do this later.

- Added a tutorial script under `source/standalone/tutorials/05_controllers/run_osc.py` and a tutorial documentation under `docs/source/tutorials/05_controllers/run_osc.rst` that explains the controller and the script on an example in which the robot applies hybrid control on a tilted surface. However, I didn't manage to render the .rst file correctly on my side, so there might be problems with the format.

From my side, the PR is ready to be merged in terms of features, and I will be happy to integrate your feedback/reviews. Once it is merged, I will create other PRs for the other potential features as discussed.

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/actions/actions_cfg.py:
```diff
@@ -248,3 +248,49 @@ class OffsetCfg:
+
+    class_type: type[ActionTerm] = task_space_actions.OperationalSpaceControllerAction
+
+    joint_names: list[str] = MISSING
+    """List of joint names or regex expressions that the action will be mapped to."""
+    body_name: str = MISSING
+    """Name of the body or frame for which motion/force control is performed."""
+    body_offset: OffsetCfg | None = None
+    """Offset of target frame w.r.t. to the body frame. Defaults to None, in which case no offset is applied."""
+    task_frame_rel_path: str = None
+    """The path of a RigidObject, relative to the sub-environment, that represents the task frame. Defaults to None."""
+    controller_cfg: OperationalSpaceControllerCfg = MISSING
+    """The configuration for the operational space controller."""
+    position_scale: float = 1.0
+    """Scale factor for the position targets. Defaults to 1.0."""
+    orientation_scale: float = 1.0
+    """Scale factor for the orientation/rotation (quad for pose_abs and or axis-angle for pose_rel) targets. Defaults to 1.0."""
+    wrench_scale: float = 1.0
+    """Scale factor for the wrench targets. Defaults to 1.0."""
+    stiffness_scale: float = 1.0
+    """Scale factor for the stiffness commands. Defaults to 1.0."""
+    damping_ratio_scale: float = 1.0
+    """Scale factor for the damping ratio commands. Defaults to 1.0."""
```
```suggestion
    joint_names: list[str] = MISSING
    """List of joint names or regex expressions that the action will be mapped to."""
    
    body_name: str = MISSING
     """Name of the body or frame for which motion/force control is performed."""
     
    body_offset: OffsetCfg | None = None
    """Offset of target frame w.r.t. to the body frame. Defaults to None, in which case no offset is applied."""
    
    task_frame_rel_path: str = None
    """The path of a RigidObject, relative to the sub-environment, that represents the task frame. Defaults to None."""
    
    controller_cfg: OperationalSpaceControllerCfg = MISSING
    """The configuration for the operational space controller."""
    
    position_scale: float = 1.0
    """Scale factor for the position targets. Defaults to 1.0."""
    
    orientation_scale: float = 1.0
    """Scale factor for the orientation/rotation (quad for pose_abs and or axis-angle for pose_rel) targets. Defaults to 1.0."""
    
    wrench_scale: float = 1.0
    """Scale factor for the wrench targets. Defaults to 1.0."""
    
    stiffness_scale: float = 1.0
    """Scale factor for the stiffness commands. Defaults to 1.0."""
    
    damping_ratio_scale: float = 1.0
    """Scale factor for the damping ratio commands. Defaults to 1.0."""
```

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/actions/task_space_actions.py:
```diff
@@ -192,3 +196,388 @@ def _compute_frame_jacobian(self):
+                self._task_frame_transformer.data.target_quat_w[:, 0, :],
+            )
+
+    def _preprocess_actions(self, actions: torch.Tensor):
```
Please add google style docstrings for arguments

ozhanozen replied:
def _preprocess_actions(self, actions: torch.Tensor):
        """Pre-processes the raw actions for operational space control.

        Args:
            actions (torch.Tensor): The raw actions for operational space control. It is a tensor of
                shape (num_envs, action_dim).
        """

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/actions/task_space_actions.py:
```diff
@@ -192,3 +196,388 @@ def _compute_frame_jacobian(self):
+        return rigid_child_prim_path
+
+    def _resolve_command_indexes(self):
+        """Resolves the indexes for the various command elements within the command tensor."""
```
please add google style docstrings for the exceptions raised

ozhanozen replied:
def _resolve_command_indexes(self):
        """Resolves the indexes for the various command elements within the command tensor.

        Raises:
            ValueError: If any command index is left unresolved.
        """

jtigue-bdai commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/actions/task_space_actions.py:
```diff
@@ -192,3 +196,388 @@ def _compute_frame_jacobian(self):
+    """
+
+    def _first_RigidObject_child_path(self):
+        """Finds the first RigidObject child under the articulation asset."""
```
please add google style docstrings for the exceptions raised

ozhanozen replied:
def _first_RigidObject_child_path(self):
        """Finds the first RigidObject child under the articulation asset.

        Raises:
            ValueError: If no child RigidObject is found under the articulation asset.

        Returns:
            str: The path to the first RigidObject child under the articulation asset.
        """

jtigue-bdai commented on source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py:
```diff
@@ -0,0 +1,864 @@
+    Helper functions
+    """
+
+    def _run_op_space_controller(
+        self,
+        robot: Articulation,
+        osc: OperationalSpaceController,
+        ee_frame_name: str,
+        arm_joint_names: list[str],
+        target_set: torch.tensor,
+    ):
```
please add docstrings for this function

ozhanozen replied:
def _run_op_space_controller(
        self,
        robot: Articulation,
        osc: OperationalSpaceController,
        ee_frame_name: str,
        arm_joint_names: list[str],
        target_set: torch.tensor,
    ):
        """Run the operational space controller with the given parameters.

        Args:
            robot (Articulation): The robot to control.
            osc (OperationalSpaceController): The operational space controller.
            ee_frame_name (str): The name of the end-effector frame.
            arm_joint_names (list[str]): The names of the arm joints.
            target_set (torch.tensor): The target set to track.
        """

jtigue-bdai commented on source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py:
```diff
@@ -0,0 +1,864 @@
+
+        return jacobian_b, mass_matrix, gravity, ee_pose_b, ee_vel_b, root_pose_w, ee_pose_w, ee_force_b
+
+    def _update_target(
+        self,
+        osc: OperationalSpaceController,
+        root_pose_w: torch.tensor,
+        ee_pose_b: torch.tensor,
+        target_set: torch.tensor,
+        current_goal_idx: int,
+    ):
```
please add docstrings for this function

ozhanozen replied:
"""Update the target for the operational space controller.

        Args:
            osc (OperationalSpaceController): The operational space controller.
            root_pose_w (torch.tensor): The root pose in the world frame.
            ee_pose_b (torch.tensor): The end-effector pose in the body frame.
            target_set (torch.tensor): The target set to track.
            current_goal_idx (int): The current goal index.

        Returns:
            command (torch.tensor): The target command.
            ee_target_pose_b (torch.tensor): The end-effector target pose in the body frame.
            ee_target_pose_w (torch.tensor): The end-effector target pose in the world frame.
            next_goal_idx (int): The next goal index.

        Raises:
            ValueError: If the target type is undefined.
        """

jtigue-bdai commented on source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py:
```diff
@@ -0,0 +1,864 @@
+
+        return command, ee_target_pose_b, ee_target_pose_w, next_goal_idx
+
+    def _convert_to_task_frame(
+        self, osc: OperationalSpaceController, command: torch.tensor, ee_target_pose_b: torch.tensor
+    ):
+        command = command.clone()
```
please add docstrings for this function

ozhanozen replied:
"""Convert the target command to the task frame if required.

        Args:
            osc (OperationalSpaceController): The operational space controller.
            command (torch.tensor): The target command to convert.
            ee_target_pose_b (torch.tensor): The end-effector target pose in the body frame.

        Returns:
            command (torch.tensor): The converted target command.
            task_frame_pose_b (torch.tensor): The task frame pose in the body frame.

        Raises:
            ValueError: If the frame is invalid.
        """

jtigue-bdai commented on source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py:
```diff
@@ -0,0 +1,864 @@
+
+        return command, task_frame_pose_b
+
+    def _check_convergence(
+        self,
+        osc: OperationalSpaceController,
+        ee_pose_b: torch.tensor,
+        ee_target_pose_b: torch.tensor,
+        ee_force_b: torch.tensor,
+        ee_target_b: torch.tensor,
+    ):
```
please add docstrings for this function

ozhanozen replied:
"""Check the convergence to the target.

        Args:
            osc (OperationalSpaceController): The operational space controller.
            ee_pose_b (torch.tensor): The end-effector pose in the body frame.
            ee_target_pose_b (torch.tensor): The end-effector target pose in the body frame.
            ee_force_b (torch.tensor): The end-effector force in the body frame.
            ee_target_b (torch.tensor): The end-effector target in the body frame.

        Raises:
            AssertionError: If the convergence is not achieved.
            ValueError: If the target type is undefined.
        """

jtigue-bdai commented on source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py:
```diff
@@ -0,0 +1,864 @@
+            # update buffers
+            robot.update(sim_dt)
+
+    def _update_states(
+        self,
+        robot: Articulation,
+        ee_frame_idx: int,
+        arm_joint_ids: list[int],
+    ):
```
please add docstrings for this function

ozhanozen replied:
"""Update the states of the robot and obtain the relevant quantities for the operational space controller.

        Args:
            robot (Articulation): The robot to control.
            ee_frame_idx (int): The index of the end-effector frame.
            arm_joint_ids (list[int]): The indices of the arm joints.

        Returns:
            jacobian_b (torch.tensor): The Jacobian in the root frame.
            mass_matrix (torch.tensor): The mass matrix.
            gravity (torch.tensor): The gravity vector.
            ee_pose_b (torch.tensor): The end-effector pose in the root frame.
            ee_vel_b (torch.tensor): The end-effector velocity in the root frame.
            root_pose_w (torch.tensor): The root pose in the world frame.
            ee_pose_w (torch.tensor): The end-effector pose in the world frame.
            ee_force_b (torch.tensor): The end-effector force in the root frame.
        """

jtigue-bdai reviewed (commented):
@OOmotuyi  this looks great. Just a few docstring/formatting stuff. If you are ready it seems like you can pull it out of draft and make it ready for review from the wider team.

ozhanozen commented:
> @OOmotuyi this looks great. Just a few docstring/formatting stuff. If you are ready it seems like you can pull it out of draft and make it ready for review from the wider team.

@jtigue-bdai, Thanks a lot! I have addressed all your comments, updated the changelog, and marked the PR as ready for review.

ozhanozen commented:
Hi @jtigue-bdai,

I’ve addressed the comments (primarily regarding docstrings) currently marked as unresolved. Would it be okay if I go ahead and mark them as resolved?

I’ve also merged the latest changes from main and updated the relevant documentation. The CHANGELOG entries remain as [unreleased], assuming the version/date will be finalized when merging to main.

On a separate branch, I’ve implemented the nullspace controller feature for OSC and successfully deployed a trained policy on a real robot to test it. It’s working well, and I observed it significantly improves learning. I’ll create a new PR for it once this one is merged.

Thanks again for your review and feedback! 😊

kellyguo11 commented:
Thanks a lot for the addition! Could you please help update the image from .png to .jpg (jpg takes up less storage), and add the new environment to the docs (https://isaac-sim.github.io/IsaacLab/main/source/overview/environments.html#single-agent and https://isaac-sim.github.io/IsaacLab/main/source/overview/environments.html#comprehensive-list-of-environments)?

Also looks like it'll need another rebase. thanks!

kellyguo11 commented on source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/manipulation/reach/config/franka/__init__.py:
```diff
@@ -65,3 +65,27 @@
+)
+
+gym.register(
+    id="Isaac-Reach-Franka-Play-OSC-v0",
```
```suggestion
    id="Isaac-Reach-Franka-OSC-Play-v0",
```
